### PR TITLE
feat(metrics): add a basic auth

### PR DESCRIPTION
**Pull request by aurrelhebert**

### CODE REVIEW

File: middlewares/metrics_auth.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+// BasicAuth is a middleware that check Metrics endpoint basic auth and deny access when auth is invalid
+func BasicAuth() echo.MiddlewareFunc {
+	username := viper.GetString("metrics.basicauth.user")
+	password := viper.GetString("metrics.basicauth.password")
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(ctx echo.Context) error {
+			logger := factories.Logger(ctx)
+			basic_user, basic_pwd, ok := ctx.Request().BasicAuth()
+			if !ok {
+				logger.Warn("Error parsing /metrics basic auth")
```
Comment (wilfriedroset) [inline middlewares/metrics_auth.go:20]: noob question: could we be more explicit about the not ok (could be in debug)?

File: cmd/root.go [hunk 2]
```diff
@@ -74,6 +74,7 @@ func initConfig() {
 	viper.SetDefault("timeunit", "us")
 	viper.SetDefault("prometheus.fillprevious.period", "5 m")
 	viper.SetDefault("metrics.enabled", true)
+	viper.SetDefault("metrics.basicauth.enable", false)
```
Comment (wilfriedroset) [inline cmd/root.go:77]: why not `metrics.basicauth.enabled` in order to be consistent with `metrics.enabled`

File: middlewares/metrics_auth.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+	"github.com/labstack/echo/v4"
+	"github.com/ovh/erlenmeyer/factories"
+	"github.com/spf13/viper"
+)
+
+// BasicAuth is a middleware that check Metrics endpoint basic auth and deny access when auth is invalid
+func BasicAuth() echo.MiddlewareFunc {
+	username := viper.GetString("metrics.basicauth.user")
+	password := viper.GetString("metrics.basicauth.password")
```
Comment (Alkorin) [inline middlewares/metrics_auth.go:14]: Could the password be hashed ?

**Comment by Alkorin** [comment]
Can't you use https://echo.labstack.com/middleware/basic-auth/ ?

**Comment by aurrelhebert** [comment]
Hi thanks all for the review, PR updated to use echo middleware and an Hash password

**Comment by aurrelhebert** [comment]
Up

**Pull request merged**